src/panels/MainPanel.tsx

```diff
@@ -80,38 +99,56 @@ export function MainPanelPane({
   }));
 
   return (
-    <div className="h-full grid-cols-1 grid-rows-[auto_1fr]">
+    <div className="h-full grid-cols-1 grid-rows-[auto_1fr]" onClick={onPaneFocused} onFocus={onPaneFocused}>
       <TabPane items={items} value={activeFile?.path} onClick={handleTabClick} onCloseClick={handleTabCloseClick} />
       <div className="flex h-full w-full overflow-hidden">
-        <MainPaneViewContent activeFile={activeFile} editorRef={editorRef} pdfViewerRef={pdfViewerRef} />
+        {activeFile && activeFileAtom ? (
+          <MainPaneViewContent
+            activeFileAtom={activeFileAtom}
+            editorRef={editorRef}
+            fileId={activeFile.path}
+            pdfViewerRef={pdfViewerRef}
+          />
+        ) : (
+          <EmptyView />
+        )}
       </div>
     </div>
   );
 }
 
 interface MainPaneViewContentProps {
-  activeFile?: FileEntry;
+  activeFileAtom: Atom<Loadable<Promise<FileContent>>>;
   editorRef: React.MutableRefObject<EditorAPI | null>;
+  fileId: FileId;
   pdfViewerRef: React.MutableRefObject<PdfViewerAPI | null>;
 }
 
-export function MainPaneViewContent({ activeFile, editorRef, pdfViewerRef }: MainPaneViewContentProps) {
-  if (!activeFile || activeFile.isFolder) {
-    return <EmptyView />;
+export function MainPaneViewContent({ activeFileAtom, editorRef, pdfViewerRef }: MainPaneViewContentProps) {
+  const loadableFileContent = useAtomValue(activeFileAtom);
+
+  if (loadableFileContent.state === 'loading') {
+    return <Spinner />;
+  }
+
+  if (loadableFileContent.state === 'hasError') {
+    return <strong>Error: {String(loadableFileContent.error)}</strong>;
   }
 
-  if (activeFile.fileExtension === 'xml') {
-    return <TextView file={activeFile} />;
+  if (loadableFileContent.data.type === 'XML') {
+    return <TextView file={loadableFileContent.data} />;
   }
 
-  if (activeFile.fileExtension === 'json') {
-    return <TextView file={activeFile} textFormatter={(input) => JSON.stringify(JSON.parse(input), null, 2)} />;
+  if (loadableFileContent.data.type === 'JSON') {
+    return (
+      <TextView file={loadableFileContent.data} textFormatter={(input) => JSON.stringify(JSON.parse(input), null, 2)} />
+    );
   }
 
-  if (activeFile.fileExtension === 'pdf') {
-    return <PdfViewer file={activeFile} pdfViewerRef={pdfViewerRef} />;
+  if (loadableFileContent.data.type === 'PDF') {
+    return <PdfViewer file={loadableFileContent.data} pdfViewerRef={pdfViewerRef} />;
   }
 
   // Use TipTap editor by default!
```

Comment:
Since the tiptap default is applied when the file is loaded now, you can switch this to an exhaustiveness check. The advantage of this is that if we add a new file type in `FileContent`, TypeScript will force us to add a new case here. Because of our eslint rules, this also pushes us to use a `switch` / `case` instead of `if` statements, which I think is fine (see https://github.com/typescript-eslint/typescript-eslint/issues/4682).

```ts
  const { data } = loadableFileContent;
  switch (data.type) {
    case 'XML':
      return <TextView file={data} />;
    case 'JSON':
      return <TextView file={data} textFormatter={(input) => JSON.stringify(JSON.parse(input), null, 2)} />;
    case 'PDF':
      return <PdfViewer file={data} pdfViewerRef={pdfViewerRef} />;
    case 'TIPTAP':
      return <TipTapView editorRef={editorRef} file={data} />;
    default:
      assertNever(data);
  }
```

Here's an implementation of `assertNever`:

```ts
/** Use this to assert that a branch is unreachable (for exhaustiveness checking). */
function assertNever(val: never) {
  throw new Error(`Reached the unreachable: ${val}`);
}
```